test/extensions/filters/http/admission_control/admission_control_filter_test.cc

```diff
@@ -0,0 +1,249 @@
+
+#include "common/stats/isolated_store_impl.h"
+
+#include "extensions/filters/http/admission_control/admission_control.h"
+#include "extensions/filters/http/admission_control/evaluators/response_evaluator.h"
+#include "extensions/filters/http/admission_control/thread_local_controller.h"
+
+#include "test/mocks/runtime/mocks.h"
+#include "test/mocks/server/mocks.h"
+#include "test/mocks/thread_local/mocks.h"
+#include "test/test_common/simulated_time_system.h"
+#include "test/test_common/utility.h"
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::NiceMock;
+using testing::Return;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace AdmissionControl {
+namespace {
+
+class MockThreadLocalController : public ThreadLocal::ThreadLocalObject,
+                                  public ThreadLocalController {
+public:
+  MockThreadLocalController() = default;
+  MOCK_METHOD(uint32_t, requestTotalCount, (), (override));
+  MOCK_METHOD(uint32_t, requestSuccessCount, (), (override));
+  MOCK_METHOD(void, recordSuccess, (), (override));
+  MOCK_METHOD(void, recordFailure, (), (override));
+};
+
+class MockResponseEvaluator : public ResponseEvaluator {
+public:
+  MockResponseEvaluator() = default;
+  MOCK_METHOD(bool, isHttpSuccess, (uint64_t code), (const, override));
+  MOCK_METHOD(bool, isGrpcSuccess, (uint32_t status), (const, override));
+};
+
+class TestConfig : public AdmissionControlFilterConfig {
+public:
+  TestConfig(const AdmissionControlProto& proto_config, Runtime::Loader& runtime,
+             TimeSource& time_source, Runtime::RandomGenerator& random, Stats::Scope& scope,
+             ThreadLocal::SlotPtr&& tls, MockThreadLocalController& controller,
+             std::shared_ptr<ResponseEvaluator> evaluator)
+      : AdmissionControlFilterConfig(proto_config, runtime, time_source, random, scope,
+                                     std::move(tls), evaluator),
+        controller_(controller) {}
+  ThreadLocalController& getController() const override { return controller_; }
+
+private:
+  MockThreadLocalController& controller_;
+};
+
+class AdmissionControlTest : public testing::Test {
+public:
+  AdmissionControlTest() = default;
+
+  std::shared_ptr<AdmissionControlFilterConfig> makeConfig(const std::string& yaml) {
+    AdmissionControlProto proto;
+    TestUtility::loadFromYamlAndValidate(yaml, proto);
+    auto tls = context_.threadLocal().allocateSlot();
+    evaluator_ = std::make_shared<MockResponseEvaluator>();
+
+    return std::make_shared<TestConfig>(proto, runtime_, time_system_, random_, scope_,
+                                        std::move(tls), controller_, evaluator_);
+  }
+
+  void setupFilter(std::shared_ptr<AdmissionControlFilterConfig> config) {
+    filter_ = std::make_shared<AdmissionControlFilter>(config, "test_prefix.");
+    filter_->setDecoderFilterCallbacks(decoder_callbacks_);
+  }
+
+  void sampleGrpcRequest(std::string&& grpc_status) {
```

Comment:
doesnt really matter a whole lot but i would just have these either be taken by value or by const-ref. forcing these to be rvalue references doesn't add anything here, especially since we end up copying the string on L83 anyways.

same for `sampleHttpRequest`

Reply:
alternatively you could have these helpers take the enum values, which might end up being a bit more readable